lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+                                       mlir::MLIRContext &DestinationContext) {
+  llvm::SmallString<1024> Buffer;
+  llvm::raw_svector_ostream Out(Buffer);
+  mlir::writeBytecodeToFile(SourceModule, Out);
+
+  mlir::Block OuterBlock;
+  mlir::LogicalResult
+    R = mlir::readBytecodeFile(llvm::MemoryBufferRef(Buffer, "cloneModuleInto"),
+                               &OuterBlock,
+                               mlir::ParserConfig(&DestinationContext));
+  revng_assert(mlir::succeeded(R));
+
+  auto &Operations = OuterBlock.getOperations();
+  revng_assert(Operations.size() == 1);
+
+  auto DestinationModule = mlir::cast<mlir::ModuleOp>(Operations.front());
+  DestinationModule->remove();
+
+  const size_t SrcOpsCount = getModuleOperations(SourceModule).size();
+  const size_t NewOpsCount = getModuleOperations(DestinationModule).size();
+  revng_assert(NewOpsCount == SrcOpsCount);
+
+  return DestinationModule;
+}
+
+static pipeline::Target makeTarget(const llvm::StringRef Name) {
+  return pipeline::Target(Name, TheKind);
+}
+
+static void makeExternal(mlir::FunctionOpInterface F) {
+  revng_assert(F->getNumRegions() == 1);
+  // A function is made external by clearing its region.
+  F->getRegion(0).getBlocks().clear();
+}
+
+// Remove any unused non-target symbols.
+static void pruneSymbols(mlir::ModuleOp Module) {
+  llvm::DenseSet<mlir::Operation *> UsedSymbols;
+
+  visit(Module, [&](mlir::FunctionOpInterface F) {
+    if (getFunctionName(F))
+      UsedSymbols.insert(F);
+
+    if (F.isExternal())
+      return;
+
+    if (const auto &Uses = mlir::SymbolTable::getSymbolUses(F)) {
+      for (const mlir::SymbolTable::SymbolUse &Use : *Uses) {
+        const auto &SymbolRef = Use.getSymbolRef();
+        revng_assert(SymbolRef.getNestedReferences().empty());
+
+        mlir::Operation *const
+          Symbol = mlir::SymbolTable::lookupSymbolIn(Module,
+                                                     SymbolRef
+                                                       .getRootReference());
+        revng_assert(Symbol);
+
+        UsedSymbols.insert(Symbol);
+      }
+    }
+  });
+
+  visit(Module, [&](mlir::SymbolOpInterface S) {
+    if (not UsedSymbols.contains(S.getOperation()))
+      S->erase();
+  });
+}
+
+const char MLIRContainer::ID = 0;
+
+void MLIRContainer::setModule(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule) {
+  revng_assert(NewModule);
+
+  // Make any non-target functions external.
+  visit(*NewModule, [&](mlir::FunctionOpInterface F) {
+    if (not F.isExternal() and not getFunctionName(F))
+      makeExternal(F);
+  });
+
+  pruneSymbols(*NewModule);
+  Module = std::move(NewModule);
+}
+
+std::unique_ptr<pipeline::ContainerBase>
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto NewContainer = std::make_unique<MLIRContainer>(name());
+  if (not getModuleBlock(*Module).empty()) {
+    OwningModuleRef NewModule(mlir::cast<mlir::ModuleOp>((*Module)->clone()));
```

Comment:
> Not that

Is this a typo? Did you mean "Not that" or "Note that"? In any case, the reason for doing the filtering in the source module is mainly to avoid dragging unused types and attributes into the new context. I'll add an explanation of that.

Reply:
It was a "note that", yup.